### ISSUE TITLE
Python return types in 4.10 break type checking due to not depending on input types

### ISSUE THREAD

**Issue by johnthagen** [opencv/opencv#25713]
### System Information

- opencv-python-headless 4.10.0.82
- numpy 1.26.4
- mypy 1.10.0
- MacOS 14.5
- Python 3.12.2

### Detailed description

Many return types in common OpenCV functions have a return type of `cv2.typing.MatLike`. The problem is that when an array with a specific `dtype` is passed in, the resulting output type does not consider this and instead returns a looser type.

### Steps to reproduce

As one example, consider `cv2.putText`. 

> Note: This is just one example, but the underlying problem is pervasive across many OpenCV Python function return type annotations.

```py
import numpy as np
import cv2

image = np.zeros((100, 100, 3), dtype=np.uint8)
image = cv2.putText(image, "Text", (20, 20), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0), 1)
```

Running Mypy:

```
$ mypy main.py      
main.py:5: error: Incompatible types in assignment (expression has type "ndarray[Any, dtype[integer[Any] | floating[Any]]]", variable has type "ndarray[Any, dtype[unsignedinteger[_8Bit]]]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

This is blocking us from upgrading from OpenCV 4.8 to 4.10 (upgrading to 4.9 was similarly blocked on other type checking problems):

- #24803

We cannot update to OpenCV 4.10 because every time we pass in a `uint8` dtype array into OpenCV, the type checker loses this specific type on the return of the OpenCV function, which generates lots of downstream type checking errors when we pass these results into other functions or dataclasses where we restrict the type annotation to a specific `dtype` (e.g., `ndarray[Any, np.uint8]`).

Mypy configuration:

```toml
[tool.mypy]
strict = true
plugins = [
    "numpy.typing.mypy_plugin",
]
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by johnthagen** [comment]
There was a previous suggestion that perhaps `TypeVar` could help with this?

- https://github.com/opencv/opencv/issues/24803#issuecomment-1874982089

CCing @VadimLevin who worked on the last type checking improvements.

**Comment by VadimLevin** [comment]
These can not be done in automatic manner (Type hints are generated as a part of build procedure) until someone create a list of all functions preserving  input/output data types of array, because there is no such information in their signature.

**Comment by asmorkalov** [comment]
cc @avokin2

**Comment by johnthagen** [comment]
> until someone create a list of all functions preserving input/output data types of array

@VadimLevin @asmorkalov Here is a list of functions that return the same `dtype` type as the source:

- `cv2.bitwise_(and | or | not | xor)`
- `cv2.approxPolyDP`
- `cv2.cvtColor`
- `cv2.morphologyEx`
- `cv2.erode`
- `cv2.dilate`
- `cv2.medianBlur`
- `cv2.gaussianBlur`
- `cv2.bilateralFilter`

Perhaps we could start here?

**Comment by johnthagen** [comment]
Related (but probably unable to fix), here a list of functions that have a `ddepth` or `ltype` input argument, who's _value_ determines the output type.

Since the type is based on a _value_ rather than another _type_, these are probably impossible to fix within the constraints of Python's current type system. But I thought I'd list them here as a reference as it's a related issue.

* `cv2.distanceTransform`
* `cv2.connectedComponents`
* `cv2.filter2D` (most all of the filtering functions)
* `cv2.addWeighted`

**Comment by johnthagen** [comment]
@devatbosch @Kumataro Since you both worked on 

- #26818

I thought you might also be interested in this issue here. It's likely a bit more involved, but I think using Generics it can be solved

- https://github.com/opencv/opencv/issues/25713#issuecomment-2150064396

**Comment by devatbosch** [comment]
Hi @johnthagen ,

Thanks for bringing it up. Sure, would like to have a look onto it...

**Comment by devatbosch** [comment]
Hi @VadimLevin ,

Just a thought, can modifying [https://github.com/opencv/opencv/blob/08a24ba2cf5007a2639035cc9661e18f3637223e/modules/python/src2/typing_stubs_generation/predefined_types.py#L38-L41](url), this segment of the code in predefined_datatypes.py help to fix this issue.

Your guidance in this record would be helpful.

Thank You!

**Comment by VadimLevin** [comment]
No, the issue is trickier than 1 line change in `predefined_types.py`.
There is a particular set of functions, many thanks to @johnthagen for gathering them, preserving type of input array.
This is not true in general case, so type hints for this functions have to be manually "refined".

Unfortunately, describing all details on how to do such refinement takes the same time as doing it.

**Comment by devatbosch** [comment]
@johnthagen , sry for the delay. As @VadimLevin mentions you carefully gathered particular set of functions preserving type of input array, so would be great if you can hint me a way we can proceed to fix this issue up...

**Comment by johnthagen** [comment]
@devatbosch I'm not familiar with the Open3D code base, but the generated type hints likely need to use generics (i.e. `TypeVar`) so that type checkers can track that the output parameters have the same type as the input parameters.

I put some link to resources about that in a previous comment:

- https://github.com/opencv/opencv/issues/25713#issuecomment-2150064396
